core/contracts/financial_templates/Liquidatable.sol

```diff
@@ -61,12 +61,18 @@ contract Liquidatable is PricelessPositionManager {
          * Following variables set upon a dispute request
          */
 
-        // Person who is disputing a liquidation
+        // Address who is disputing a liquidation
         address disputer;
         // Time when dispute is initiated, needed to get price from Oracle
         uint disputeTime;
         // Final price as determined by an Oracle following a dispute
         FixedPoint.Unsigned settlementPrice;
+        /**
+         * Following variables check that each member can only withdraw once
+         */
+        bool hasSponsorWithdrawn;
+        bool hasLiquidatorWithdrawn;
+        bool hasDisputorWithdrawn;
```

Comment:
Do you think there's a way to signal that these users have withdrawn without adding these bools? Maybe by deleting their addresses or something?

Very half baked idea: instead of using liquidator to determine whether the liquidation is "valid" (when we access it), we could use tokensOutstanding, or something. Then we could just progressively delete addresses as the different users withdraw.

Not sure if that works, but it may make some of the checks below simpler and it may allow us to reduce the size/complexity of this struct.

Reply:
For sure. I agree with you 100% that there could be a more storage efficient way of achieving this. Deleting the address is actually a good solution because the check we currently have right now:
```
require(
        (msg.sender == liquidation.disputer) || (msg.sender == liquidation.liquidator) || (msg.sender == sponsor),
            "must be a disputer, liquidator, or sponsor to request a withdrawal on a liquidation"
        );
```
would revert once deleted. For me this is a trade off between explicit memory storage with a permanent history of users in different rolls vs a more efficient and implicit way of achieving the same goal.

Reply:
Generally, I think logs are a better (cheaper :) ) way of getting general purpose contract history. Especially if we don't have a use case for that history yet.

Reply:
I agree with you but if we want to remove the address from the struct to notate payment for a user we introduce some complexities. For one this works fine for the liquidator and the disputer, but not the sponsor. This is because the sponsor's address is the key in a mapping for each liquidation object and as such we cant easily remove it. we could solve this by having either:
1) a bool specifically to represent if the sponsor has withdrawn. This is clunky because it uses a different method for the sponsor, liquidator & disruptor.
2) store the sponsors address within the struct as well. this is also not ideal as we now have the sponsor address acting as both the key for the struct and a value within the struct.